runtime/compiler/optimizer/BoolArrayStoreTransformer.cpp

```diff
@@ -0,0 +1,434 @@
+      if (signature && length >= 2 && signature[length-1] == expectedTypeChar && signature[length-2] == '[')
+         nodeArrayDimension = length-1;
+      }
+	return nodeArrayDimension;
+   }
+
+static char * getTypeName(TR_YesNoMaybe type, char * buffer)
+   {
+   switch (type)
+      {
+      case TR_yes:
+         strcpy(buffer, "[Z");
+         break;
+      case TR_no:
+         strcpy(buffer, "[B");
+         break;
+      case TR_maybe:
+         strcpy(buffer, "unknown type");
+         break;
+      }
+   return buffer;
+   }
+
+void printTypeInfo(TR_BoolArrayStoreTransformer::TypeInfo *typeInfo, TR::Compilation *comp)
+   {
+   int localIndex = 0;
+   for (auto it = typeInfo->begin(); it != typeInfo->end(); it++)
+      {
+      if (*it != TR_maybe)
+         {
+         char buffer[15];
+         traceMsg(comp, "( local # %d: %s )  ", localIndex, getTypeName(*it, buffer));
+         }
+      localIndex++;
+      }
+   }
+
+/*
+ * \brief
+ *   This function caculcates the type info of each auto and figure out whether a
+ *   bstorei node is for boolean array
+ *
+ * \parm block
+ *    The block to process
+ *
+ * \parm blockStartTypeInfo
+ *    The auto type info at block start
+ *
+ * \return currentTypeInfo
+ *    The auto type info at the end of the current block
+ *
+ * \note
+ *    The function looks at each store to auto and record whether the type is boolean array (TR_yes), byte array type
+ *    (TR_no), other or unknown type (TR_maybe).
+ */
+TR_BoolArrayStoreTransformer::TypeInfo * TR_BoolArrayStoreTransformer::processBlock(TR::Block *block, TR_BoolArrayStoreTransformer::TypeInfo *blockStartTypeInfo)
+   {
+   TR_BoolArrayStoreTransformer::TypeInfo *currentTypeInfo = blockStartTypeInfo;
+   if (comp()->getOption(TR_TraceILGen))
+      {
+      traceMsg(comp(), "start processing block_%d: ", block->getNumber());
+      if (currentTypeInfo)
+         printTypeInfo(currentTypeInfo, comp());
+      traceMsg(comp(), "\n");
+      }
+   for (TR::TreeTop *tt = block->getEntry(); tt != block->getExit(); tt = tt->getNextTreeTop())
+      {
+      TR::Node *node = tt->getNode();
+      if (node->getOpCode().isStoreDirect())
+         {
+         TR::Node *rhs = node->getFirstChild();
+         TR::AutomaticSymbol *local = node->getSymbolReference()->getSymbol()->getAutoSymbol();
+         if (local && rhs->getDataType().isAddress())
+            {
+            TR_YesNoMaybe newType = TR_maybe;
+            if (isBoolArrayNode(rhs))
+               newType = TR_yes;
+            else if (isByteArrayNode(rhs))
+               newType = TR_no;
+            else if (rhs->getOpCode().hasSymbolReference())
+               {
+               TR::AutomaticSymbol *rhsLocal = rhs->getSymbolReference()->getSymbol()->getAutoSymbol();
+               if (rhsLocal && currentTypeInfo)
+                  newType = (*currentTypeInfo)[rhsLocal->getLocalIndex()];
+               }
+
+            if (newType != TR_maybe || currentTypeInfo)
```

Comment:
added `|| currentTypeInfo` to catch the case where currentTypeInfo is `[Z` or `[B` but the new type is unknown.